test/support/utils_test_base.h

```diff
@@ -0,0 +1,717 @@
+{
+    eKeys = 0,  // Keys
+    eVals,      // Values
+    eRes        // Results
+};
+
+template <typename TEnum>
+auto
+enum_val_to_index(TEnum enumVal)
+    -> decltype(static_cast<typename ::std::underlying_type<TEnum>::type>(enumVal))
+{
+    return static_cast<typename ::std::underlying_type<TEnum>::type>(enumVal);
+}
+
+template <typename TestValueType>
+struct test_base_data_visitor;
+
+////////////////////////////////////////////////////////////////////////////////
+// struct test_base_data - test source data base class
+template <typename TestValueType>
+struct test_base_data
+{
+    /// Visit all test data
+    /**
+     * @param test_base_data_visitor<TestValueType>* visitor - pointer to visitor
+     */
+    virtual void visit(test_base_data_visitor<TestValueType>* visitor) = 0;
+};
+
+#if TEST_DPCPP_BACKEND_PRESENT
+////////////////////////////////////////////////////////////////////////////////
+/// struct test_base_data_usm -  test source data for USM shared/device memory
+template <typename TestValueType>
+struct test_base_data_usm : test_base_data<TestValueType>
+{
+    struct Data
+    {
+        using TSourceData = usm_data_transfer_base<TestValueType>;
+        using TSourceDataPtr = ::std::unique_ptr<TSourceData>;
+
+        sycl::usm::alloc alloc_type;        // USM alloc type (shared/device)
+        TSourceDataPtr   src_data_usm;      // USM data transfer helper
+        ::std::size_t    offset = 0;        // Offset in USM buffer
+
+        template<typename _Size>
+        Data(sycl::usm::alloc __alloc_type, sycl::queue __q, _Size __sz, ::std::size_t __offset)
+            : alloc_type(__alloc_type)
+            , offset(__offset)
+        {
+            // We use this switch/case because we have test_base_data_visitor interface with
+            // virtual functions, arguments of which can't be template classes.
+            // So, USM allocation type specified in runtime.
+            switch (__alloc_type)
+            {
+            case sycl::usm::alloc::shared:
+                src_data_usm.reset(new usm_data_transfer<sycl::usm::alloc::shared, TestValueType>(__q, __sz + __offset));
+                break;
+            case sycl::usm::alloc::device:
+                src_data_usm.reset(new usm_data_transfer<sycl::usm::alloc::device, TestValueType>(__q, __sz + __offset));
+                break;
+            default:
+                assert(false);
+                break;
+            }
+        }
+
+        auto get_usm_data_shared()
+        {
+            assert(alloc_type == sycl::usm::alloc::shared);
+            return reinterpret_cast<usm_data_transfer<sycl::usm::alloc::shared, TestValueType>*>(src_data_usm.get());
+        }
+
+        auto get_usm_data_device()
+        {
+            assert(alloc_type == sycl::usm::alloc::device);
+            return reinterpret_cast<usm_data_transfer<sycl::usm::alloc::device, TestValueType>*>(src_data_usm.get());
+        }
+
+        TestValueType* get_start_from()
```

Comment:
goal-?

Reply:
Goal - to receive pointer from base class to derived class with required template parameters.
Added comments.